packages/gdl-frontend/lib/facebookPixel.js

```diff
@@ -0,0 +1,57 @@
+// @flow
+import ReactPixel from 'react-facebook-pixel';
+import getConfig from 'next/config';
+import type { ConfigShape } from '../types';
+
+const {
+  publicRuntimeConfig: { facebookPixelId }
+}: ConfigShape = getConfig();
+
+let REACT_PIXEL_INITIALIZED = false;
+
+export function initFacebookPixel() {
+  if (facebookPixelId && !REACT_PIXEL_INITIALIZED) {
+    const options = {
+      autoConfig: true,
+      debug: true
+    };
+    ReactPixel.init(facebookPixelId, options);
+    REACT_PIXEL_INITIALIZED = true;
+  }
+}
+
+export function facebookPixelPageView() {
+  if (REACT_PIXEL_INITIALIZED) {
+    ReactPixel.pageView();
+  }
+}
+
+// Define all allowed event categories here, so we are able to group all datapoints correctly
+type EventCategory = 'PageView' | 'Search' | 'ViewContent';
+type Action =
+  // User
+  | 'Login'
+  // Books
+  | 'Book details'
+  | 'Read'
+  | 'Favorited'
+  // Navigation
+  | 'Home'
+  | 'Category'
+  | 'Language'
+  | 'Featured'
+  | 'More - Search'
+  | 'More - Browse'
+  | 'Search';
+
+//  This function makes it possible to use other events than 'PageView' to get statistics
+export function logEventToFacebookPixel(
+  category: EventCategory,
+  action: Action,
+  data: Object
+) {
+  if (REACT_PIXEL_INITIALIZED) {
+    // track(event, data) : https://developers.facebook.com/docs/facebook-pixel/implementation/conversion-tracking#pixel-standard-events
+    ReactPixel.track(category, { action: action, ...(data && data) });
```

Comment:
Since `data` is not optional and even if it was optional, the spread operator handle it so you can write `{ action, ...data }`.

Fyi, the order of using the spread operator matters. If you set the spread last, it takes priority/precedence over the other values which in this case if you have a ` data = { action: 'Event' }` it will overwrite the earlier `action`. I don't know if that is your purpose so don't bother if you knew it!

Reply:
Good catch :) As for now I don't think it will be a problem. And we may be adjusting this function in the future depending on what the customer wishes to achieve.